nautobot/core/api/schema.py

```diff
@@ -0,0 +1,250 @@
+import logging
+import re
+
+from drf_spectacular.extensions import OpenApiSerializerFieldExtension
+from drf_spectacular.openapi import AutoSchema
+from rest_framework import serializers
+from rest_framework.relations import ManyRelatedField
+
+from nautobot.core.api import (
+    ChoiceField,
+    SerializedPKRelatedField,
+    WritableNestedSerializer,
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+class NautobotAutoSchema(AutoSchema):
+    """Nautobot-specific extensions to drf-spectacular's AutoSchema."""
+
+    custom_actions = ["bulk_update", "bulk_partial_update", "bulk_destroy"]
+
+    def get_operation_id(self):
+        """Extend the base method to handle Nautobot's REST API bulk operations.
+
+        Without this extension, every one of our ModelViewSet classes will result in drf-spectacular complaining
+        about operationId collisions, e.g. between DELETE /api/dcim/devices/ and DELETE /api/dcim/devices/<pk>/ would
+        both get resolved to the same "dcim_devices_destroy" operation-id and this would make drf-spectacular complain.
+
+        With this extension, the bulk endpoints automatically get a different operation-id from the non-bulk endpoints.
+        """
+        if hasattr(self.view, "action") and self.view.action in self.custom_actions:
+            # Same basic sequence of calls as AutoSchema.get_operation_id,
+            # except we use "self.view.action" instead of "self.method_mapping[self.method]" to get the action verb
+            tokenized_path = self._tokenize_path()
+            tokenized_path = [t.replace("-", "_") for t in tokenized_path]
+
+            action = self.view.action
+
+            if not tokenized_path:
+                tokenized_path.append("root")
+            if re.search(r"<drf_format_suffix\w*:\w+>", self.path_regex):
+                tokenized_path.append("formatted")
+
+            return "_".join(tokenized_path + [action])
+
+        # For all other view actions, operation-id is the same as in the base class
+        return super().get_operation_id()
+
+    def get_request_serializer(self):
+        """Return the request serializer (used for parsing the request payload) for this endpoint.
```

Comment:
`nit`: We seem to have comments on the same line as the block comment `"""` and in other places (see below) where they start on a newline. Is there a reason for this?

Reply:
Two different valid docstring styles that we're not consistent about everywhere.  I learned with the same-line style so it's what I generally default to out of habit but I think most of nautobot uses the new-line style. 

There are linters such as `pydocstyle` that we could enable and configure to enforce one style or the other at some point.